server/src/main/java/org/apache/druid/segment/loading/SegmentLoaderLocalCacheManager.java

```diff
@@ -203,8 +219,8 @@ public Segment getSegment(DataSegment segment, boolean lazy, SegmentLazyLoadFail
     } else {
       factory = new MMappedQueryableSegmentizerFactory(indexIO);
     }
-
-    return factory.factorize(segment, segmentFiles, lazy, loadFailed);
+    Segment baseSegment = factory.factorize(segment, segmentFiles, lazy, loadFailed);
+    return ReferenceCountingSegment.wrapSegment(baseSegment, segment.getShardSpec());
```

Comment:
Since it's a stateful object with a reference count, it seems a little weird to me that each call is returning a new instance (which would have a fresh reference state)

Reply:
IMO It should return a different `ReferenceCountingSegment` here since the base object too is being re-created. That being said, other implementations can return the same object if they can. 

Ideally, I wanted to return a `SegmentReference` here but `ReferenceCountingObject` is heavily used already.

Reply:
Before, the DataSegment was an immutable object, so recreating it didn't really matter. Conceptually, there's only 1 copy of a segment on a historical, so having multiple potentially different views of how many references it has seems odd to me.